packages/@uppy/companion-client/src/RequestClient.ts

```diff
@@ -0,0 +1,613 @@
+        cause: err,
+      })
+    }
+  }
+
+  async get<PostBody extends Record<string, unknown>>(
+    path: string,
+    options?: RequestOptions,
+  ): Promise<PostBody> {
+    // TODO: remove boolean support for options that was added for backward compatibility.
+    // eslint-disable-next-line no-param-reassign
+    if (typeof options === 'boolean') options = { skipPostResponse: options }
+    return this.request({ ...options, path })
+  }
+
+  async post<PostBody extends Record<string, unknown>>(
+    path: string,
+    data: Record<string, unknown>,
+    options?: RequestOptions,
+  ): Promise<PostBody> {
+    // TODO: remove boolean support for options that was added for backward compatibility.
+    // eslint-disable-next-line no-param-reassign
+    if (typeof options === 'boolean') options = { skipPostResponse: options }
+    return this.request<PostBody>({ ...options, path, method: 'POST', data })
+  }
+
+  async delete(
+    path: string,
+    data: Record<string, unknown>,
+    options?: RequestOptions,
+  ): Promise<unknown> {
+    // TODO: remove boolean support for options that was added for backward compatibility.
+    // eslint-disable-next-line no-param-reassign
+    if (typeof options === 'boolean') options = { skipPostResponse: options }
+    return this.request({ ...options, path, method: 'DELETE', data })
+  }
+
+  /**
+   * Remote uploading consists of two steps:
+   * 1. #requestSocketToken which starts the download/upload in companion and returns a unique token for the upload.
+   * Then companion will halt the upload until:
+   * 2. #awaitRemoteFileUpload is called, which will open/ensure a websocket connection towards companion, with the
+   * previously generated token provided. It returns a promise that will resolve/reject once the file has finished
+   * uploading or is otherwise done (failed, canceled)
+   */
+  async uploadRemoteFile(
+    file: UppyFile<M, B>,
+    reqBody: Record<string, unknown>,
+    options: { signal: AbortSignal; getQueue: () => RateLimitedQueue },
+  ): Promise<void> {
+    try {
+      const { signal, getQueue } = options || {}
+
+      return await pRetry(
+        async () => {
+          // if we already have a serverToken, assume that we are resuming the existing server upload id
+          const existingServerToken = this.uppy.getFile(file.id)?.serverToken
+          if (existingServerToken != null) {
+            this.uppy.log(
+              `Connecting to exiting websocket ${existingServerToken}`,
+            )
+            return this.#awaitRemoteFileUpload({
+              file,
+              queue: getQueue(),
+              signal,
+            })
+          }
+
+          const queueRequestSocketToken = getQueue().wrapPromiseFunction(
+            async (args: {
+              file: UppyFile<M, B>
+              postBody: Record<string, unknown>
+              signal: AbortSignal
+            }) => {
```

Comment:
```suggestion
            // This should accept ReturnType<typeof this.#requestSocketTocken>
            async (...args: any) => {
```

Reply:
I tried but you can't infer types of private fields

Reply:
That's why I'm suggesting adding a comment

Reply:
I'd rather take correctly typed over `any`. Duplication is fine as long as it doesn't happen to often.

Reply:
I think `any` is safer, we don't do anything with it but passing it to the private method. If the private method changes, we might not update this, which would lead to bug. If we use any, we dodge that bullet.

Reply:
I disagree, I think spreading all the arguments is a better pattern (i.e. `...args`, or using `arguments`). It's a shame that TS has this bug/limitation, but I don't think it's a good idea to impose it onto ourselves when we could just tell TS you know better.
Another workaround suggested in the TS issue is to define the types outside the class, and apply them everywhere we need them

```typescript
type requestSocketTokenArgs = [{
              file: UppyFile<M, B>
              postBody: Record<string, unknown>
              signal: AbortSignal
}]

/* … */

async (...args: requestSocketTokenArgs) => {
  /* … */
}

/* … */

#requestSocketToken (...args: requestSocketTokenArgs) {
  const { file, postBody, signal } = args[0]
  /* … */
}

Reply:
> It's possible to do this, but I really think what we currently have is better:

I 💯 with this, if you mean "what we currently have on `main`". The problem with this implementation change is that it limits the number of arguments to 1. I think we should be passing all the arguments, so if we add more we don't need to touch those intermediary functions. If you want to duplicate the types, let's at least not touch the output code:

```suggestion
            async (...args: [{
              file: UppyFile<M, B>
              postBody: Record<string, unknown>
              signal: AbortSignal
            }]) => {
```